design-proposals/klip-8-interactive-queries.md

```diff
@@ -131,13 +131,129 @@ Additionally:
+
+* De-emphasizes the stream abstraction as we continue to consider making KSQL more table centric.
+* Not redundant in the context of CSAS (`CREATE STREAM s0 AS SELECT .. FROM s1 EMIT STREAM`).
+* Intuitive and descriptive in the context of tables. A streaming query against a table effectively produces rows representing changes to the underlying table. Identifying this changelog as simply a stream isn't necessarily clear.
+* Similarly to the above, `EMIT CHANGES` is intuitive in the context of streaming aggregations, which produce incremental *changes* over time.
+* `EMIT STREAM` generally feels better suited for ad-hoc queries, for which there is no other `STREAM` context in the query (i.e. CSAS).
+
+The following are canonical example queries within each of the aforementioned fundamental query contexts:
+
+```sql
+-- Streaming query on a stream
+SELECT x, count(*) FROM stream GROUP BY x EMIT CHANGES;
+
+-- Streaming query on a table
+SELECT x, y FROM table WHERE x < 10 EMIT CHANGES;
+
+-- Point-in-time query on a table
+SELECT ROWKEY, count FROM table WHERE ROWKEY = 'key';
+
+-- Point-in-time query on a stream (semantics unclear, illustrative example only)
+SELECT * FROM stream WHERE column = 42;
+```
+
+---
+
+Finally, `EMIT CHANGES` may be preferrable over `EMIT STREAM` for KSQL specifically (for the reasons outlined above) although `EMIT STREAM` is still a suitable option.
+
+## Proposed behavioral changes
+
+As we add this fundamentally new capability to KSQL, it is worth considering any modifications we can make to KSQL's existing behavior to ensure query behavior is as intuitive as possible after the addition of PIT queries. There is one potential behavioral change that is worth considering in this respect: **streaming queries against a table should return the table's entire materialized state (possibly filtered by a `WHERE` clause) followed by changes.**
+
+Currently, a streaming query against a table in KSQL will only return rows from the table's changelog produced *after* the query begins (assuming a default start offset of `latest`).
+
+However, streaming queries returning a table's entire materialized state *then* followed by changes is more consistent with the behavior of PIT queries on a table. This behavior is also likely what users want when issuing a streaming query against a table, and it is currently not straightforward to achieve.
+
+To get KSQL's current behavior with a streaming query against a table, users would simply use an intuitive `WHERE` clause:
+
+```sql
+SELECT * FROM table WHERE ROWTIME >= NOW EMIT CHANGES;
+```
+
+## Limitations
+
+The following are proposed intentional limitations of the initial implementation of PIT queries:
+
+**PIT queries on streams probably do not need to be initially supported.** The semantics and use cases around these are unclear and require further analysis. However, the proposed syntax leaves open the possibility of supporting PIT queries on streams without further modifications.
```

Comment:
I agree that support for PIT queries on streams can be postponed.

However, I disagree with the opinions of some in this discussion who question the value of PIT queries on streams. Think of this feature as, for example, being able to "grep" a Kafka topic like you can do `grep whatIsInteresting myFile`. This is a very convenient way to do data exploration of a Kafka topic and also useful for troubleshooting and debugging, to name but two examples.

Also, it conceptually doesn't make sense to me to evangelize the stream-table duality but then taking the view point of "We believe you users should not be allowed to do query streams and tables in the same way". I can't see any reason personally why we should intentionally decide to not (i.e., never) support PIT queries on streams. I do agree, to be clear, that it's perfectly fine to postpone support for such functionality because of LOE and timeline.

Reply:
I'm with Michael here, having previously argued to opposite ;)

Having spent a great deal of time thinking about this I'm now of the strong opinion that a 'point in time query of a stream' is nothing more than running the query over the data held in the streams Kafka topic.

For example,

```sql
SELECT a, b, c FROM x where y = z;
```

Describes a point in time query over 'x' from the beginning of time to 'now'. (the 'now' is explicit, as previously mentioned, to keep PIT queries standard SQL).

If 'x' is a stream, then we read the full history from Kafka, filter by 'y=z' and return each row that matches. As a stream is not keyed like a table, each event is treated as distinct, so is a final result and hence is returned.

Of course, some of the history of 'x' may have been lost from the topic due to eviction. But we have not control over that. We can only operate on the data we have available.

There is virtually no difference between 'x' being a stream or a table.  If 'x' was a table we'd do the same thing: we'd read the changelog, apply the filter, and return the result.  The only difference is that a change log is _keyed_ and we'd therefore allow any later event to overwrite earlier ones, (e.g. if compaction hadn't run yet), and we'd handle null values as deletes. Finally, because its a table and there is no `EMIT CHANGES` we'd only return the final result: i.e. we'd read the entire changelog, materialize, and then return the final table state.

Of course, there's the possibility that the table has already been materialised in KSQL. In such a case we can server the request from the materialized state store - but this is only an optimisation: the underlying semantics remain the same.

As for whether we should support stream PIT queries initially... I say 'why not'?  If we don't, then users should have to do the `SET 'auto.offset.reset' = 'earliest'` trick, following by a streaming query, and then terminate it once its got to 'now'.  Why not just support PIT queries on streams? (Assuming time allows).

Reply:
> This is a very convenient way to do data exploration of a Kafka topic and also useful for troubleshooting and debugging, to name but two examples.

I think there are actual two distinct use cases here. a) tailing a log and greping it for a pattern continuously. the current streaming queries over streams can do that. You are spot on that this is very common.

b) then comes the full blown data exploration which can potentially involve adhoc queries that do random access or scan a whole lot of the kafka cluster.. This is not something I have seen people do very successfully with Kafka for good reasons. For the query to return in any reasonable amount of time, you need some kind of indexing like we have on tables.

Even as this blog https://www.confluent.io/blog/okay-store-data-apache-kafka/ points out, 

> "First, databases are mostly about queries, and I don’t think Kafka really benefits from trying to add any kind of random access lookups directly against the log."

 I don't think Kafka is the right system to be writing such adhoc queries against. Sure, you can make it work functionally, but would the user be happy with the "interactiveness" of the query? then are we saynig PIT queries on kafka steams will be non-interactive?